xquant/keras/keras_report_utils.py

```diff
@@ -0,0 +1,380 @@
+import numpy as np
+from keras import Model
+import tensorflow as tf
+from functools import partial
+
+from xquant import XQuantConfig
+from xquant.common.constants import INTERMEDIATE_METRICS_REPR, INTERMEDIATE_METRICS_VAL, XQUANT_REPR, XQUANT_VAL
+from xquant.common.framework_report_utils import FrameworkReportUtils, MSE_METRIC_NAME, CS_METRIC_NAME, SQNR_METRIC_NAME
+from model_compression_toolkit.ptq.keras.quantization_facade import DEFAULT_KERAS_TPC
+from model_compression_toolkit.core.keras.reader.reader import model_reader
+from xquant.logger import Logger
+
+
+class KerasReportUtils(FrameworkReportUtils):
+
+    def __init__(self, report_dir: str):
+        """
+        Args:
+            report_dir: Logging dir path.
+        """
+        tb_writer = TensorboardWriter(report_dir, DEFAULT_KERAS_INFO)
+        super().__init__(tb_writer=tb_writer)
+
+    def get_metric_on_output(self,
+                             float_model: keras.Model,
+                             quantized_model: keras.Model,
+                             dataset: Callable,
+                             custom_metrics_output: Dict[str, Callable] = None,
+                             is_validation: bool = False) -> Dict[str, float]:
+        """
+        Compute metrics on the output of the model.
+
+        Args:
+            float_model (keras.Model): The floating-point Keras model.
+            quantized_model (keras.Model): The quantized Keras model.
+            dataset (Callable): Dataset used for inference.
+            custom_metrics_output (Dict[str, Callable], optional): Custom metrics for output evaluation. Defaults to
+            None.
+            is_validation (bool, optional): Flag indicating if this is a validation dataset. Defaults to False.
+
+        Returns:
+            Dict[str, float]: A dictionary of computed metrics.
+        """
+
+        dataset = partial(self.wrapped_dataset,
+                          dataset=dataset,
+                          is_validation=is_validation)
+
+        metrics_to_compute = list(self.get_default_metrics().keys())
+        if custom_metrics_output:
+            if not isinstance(custom_metrics_output, dict):
+                Logger.critical(f"custom_metrics_output should be a dictionary but is {type(custom_metrics_output)}")
+            metrics_to_compute += list(custom_metrics_output.keys())
+
+        metrics = {key: [] for key in metrics_to_compute}
+
+        for x in dataset():
+            float_predictions = float_model.predict(x)
+            quant_predictions = quantized_model.predict(x)
+            results = self.compute_metrics((float_predictions, quant_predictions),
+                                           custom_metrics_output)
+
+            # Accumulating results
+            for key in metrics:
+                metrics[key].append(results[key])
+
+        # Averaging metrics across the dataset
+        aggregated_metrics = {key: sum(value) / len(value) for key, value in metrics.items()}
```

Comment:
is the dataset returning batches? single images? a batch for each input of the model?
anyway, I'm not sure this average with the len(value) is well-defined

Reply:
It returns a list of batches (the list's length is the number of model's inputs).
Why is it not well-defined? We average over all batches from the dataset.